File: data-structures-sorting-algorithms-advanced/src/algorithms/heapSort.js
```javascript
/**
 * @fileoverview Implementation of the Heap Sort algorithm.
 */

const { swap } = require('../utils/arrayUtils');

/**
 * Helper function to heapify a subtree rooted with node 'i'.
 * This assumes that the subtrees at 'left' and 'right' are already heapified.
 *
 * @param {Array<number>} arr The array representing the heap.
 * @param {number} n The size of the heap (number of elements in the current array segment).
 * @param {number} i The index of the root node of the subtree to heapify.
 * @returns {void} The array is modified in-place.
 */
function heapify(arr, n, i) {
  let largest = i; // Initialize largest as root
  const left = 2 * i + 1; // Left child
  const right = 2 * i + 2; // Right child

  // If left child is larger than root
  if (left < n && arr[left] > arr[largest]) {
    largest = left;
  }

  // If right child is larger than largest so far
  if (right < n && arr[right] > arr[largest]) {
    largest = right;
  }

  // If largest is not root
  if (largest !== i) {
    swap(arr, i, largest);
    // Recursively heapify the affected sub-tree
    heapify(arr, n, largest);
  }
}

/**
 * Sorts an array of numbers using the Heap Sort algorithm.
 *
 * Heap Sort is a comparison-based sorting algorithm. It's conceptually similar to selection sort,
 * but instead of scanning the entire unsorted part of the array to find the maximum element,
 * it uses a heap data structure to efficiently find the maximum.
 *
 * The algorithm involves two main steps:
 * 1. Build a max-heap from the input data.
 * 2. One by one, extract the maximum element from the heap (which is at the root),
 *    and replace it with the last element of the heap, then reduce the heap size
 *    and heapify the root. Repeat until the heap is empty.
 *
 * Characteristics:
 * - Not stable: Yes (does not preserve the relative order of equal elements).
 * - In-place: Yes.
 * - Not adaptive: Performance is consistent regardless of initial order.
 * - Guaranteed O(N log N) worst-case performance.
 *
 * @param {Array<number>} arr The array of numbers to be sorted.
 * @returns {Array<number>} The sorted array. The original array is modified in-place.
 *
 * Time Complexity:
 *   - Worst-case: O(N log N)
 *   - Average-case: O(N log N)
 *   - Best-case: O(N log N) (even on already sorted data)
 *
 * Space Complexity:
 *   - O(1) (in-place)
 *
 * Example:
 *   let arr = [5, 1, 4, 2, 8];
 *   heapSort(arr); // arr becomes [1, 2, 4, 5, 8]
 *   heapSort([]);   // returns []
 *   heapSort([1]);  // returns [1]
 */
function heapSort(arr) {
  if (!Array.isArray(arr)) {
    throw new Error('Input must be an array.');
  }
  const n = arr.length;

  // An array with 0 or 1 element is already sorted.
  if (n <= 1) {
    return arr;
  }

  // Build a max-heap (rearrange array)
  // We start from the last non-leaf node and go up to the root.
  // The index of the last non-leaf node is floor(n/2) - 1.
  for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
    heapify(arr, n, i);
  }

  // One by one extract elements from the heap
  for (let i = n - 1; i > 0; i--) {
    // Move current root to end
    swap(arr, 0, i);

    // Call max heapify on the reduced heap
    heapify(arr, i, 0);
  }
  return arr;
}

module.exports = heapSort;
```

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms_project/algorithms/heap_sort.py
def heap_sort(arr):
    """
    Sorts an array using the Heap Sort algorithm.

    Heap Sort is a comparison-based sorting algorithm that uses a binary heap data structure.
    It's an in-place sorting algorithm and has a time complexity of O(n log n) in all cases.

    The algorithm consists of two main parts:
    1. Build a max-heap from the input data.
    2. One by one, extract elements from the heap. An element is extracted by
       swapping the root of the heap (largest element) with the last element
       of the heap, then reducing the size of the heap by one, and finally
       heapifying the root.

    Args:
        arr (list): The list of comparable elements to be sorted.

    Time Complexity:
        - Worst Case: O(n log n)
        - Average Case: O(n log n)
        - Best Case: O(n log n)

    Space Complexity:
        - O(1) - It sorts the array in-place, requiring no additional storage
                 proportional to the input size.

    Stability:
        - Unstable: It does not preserve the relative order of equal elements
                    due to the swapping operations during heapify.
    """
    n = len(arr)

    # 1. Build a max-heap (rearrange array)
    # The loop runs from the last non-leaf node up to the root.
    # The index of the last non-leaf node is n // 2 - 1.
    for i in range(n // 2 - 1, -1, -1):
        _heapify(arr, n, i)

    # 2. Extract elements one by one from the heap
    for i in range(n - 1, 0, -1):
        # Move current root (largest element) to end of array
        arr[i], arr[0] = arr[0], arr[i]
        # Call max heapify on the reduced heap
        _heapify(arr, i, 0) # i is the new heap_size
    # The function modifies the array in-place.


def _heapify(arr, n, i):
    """
    A helper function to heapify a subtree rooted with node 'i' which is
    an index in arr[]. 'n' is the size of the heap.

    Args:
        arr (list): The array representing the heap.
        n (int): The current size of the heap.
        i (int): The index of the root of the subtree to heapify.
    """
    largest = i  # Initialize largest as root
    left_child = 2 * i + 1  # Left child index
    right_child = 2 * i + 2  # Right child index

    # See if left child of root exists and is greater than root
    if left_child < n and arr[left_child] > arr[largest]:
        largest = left_child

    # See if right child of root exists and is greater than current largest
    if right_child < n and arr[right_child] > arr[largest]:
        largest = right_child

    # Change root, if needed
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap
        # Heapify the root.
        _heapify(arr, n, largest)
```